# This is for future Improvement 

import warnings
#import SearchEngine as SE
warnings.filterwarnings("ignore")
import csv
import os
import pandas as pd

In [ ]:
def step_by_step_tasks(self,user_input):
        from transformers import pipeline

        summarizer = pipeline("summarization", model="facebook/bart-large-cnn")

        #ARTICLE ='i need recommendations for iron man book and select the top 1 book then add it to my favourite list after that add top 2nd 3rd books to the to_read list if 2nd 3rd top books are not in to_read list if 2nd 3rd books are already in to_read list then add 4th 5th top books to to_read list'
        """ i need recommendations for iron man book and select the top 1 book then add it to my 
        favourite list after that add top 2nd 3rd books to the to_read list if 2nd 3rd top books are not in 
        to_read list if 2nd 3rd books are already in to_read list then add 4th 5th top books to to_read list
        """
        data = summarizer(user_input, max_length=len(user_input)/3, min_length=3, do_sample=False)
        print(data)
        summary_text = data[0]['summary_text']

        # Split the text into multiple steps based on the dot ('.') character
        steps = summary_text.split('.')

        # Remove any leading or trailing whitespaces from each step
        steps = [step.strip() for step in steps if step.strip()]

        print("\n",steps)
        print("\n",steps[0])
        book_title = self.process_user_input_title_extraction(steps[0])

In [3]:
import SearchEngine as SE

In [228]:
import sys

In [4]:
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk import pos_tag, ne_chunk

import nltk
from nltk.chat.util import Chat, reflections

reflections = {
    "i am": "you are",
    "i was": "you were",
    "i": "you",
    "i'm": "you are",
    "i'd": "you would",
    "i've": "you have",
    "i'll": "you will",
    "my": "your",
    "you are": "I am",
    "you were": "I was",
    "you've": "I have",
    "you'll": "I will",
    "your": "my",
    "yours": "mine",
    "you": "me",
    "me": "you"
}

pairs = [
    ["my name is (.*)", ["Hello %1, how are you today?"]],
    ["myself (.*)", ["Hello %1, how are you today?"]],
    ["hi|hey|hello", ["Hello", "Hey there"]],
    ["(.*) special (.*)|(.*) Speciality|good at what", ["I can find recommendations for you, and i can do multiple tasks one by one but not all at once"]],
    ["(.*) future plans|future plans", ["Mr Bharath Working on it,as of now i'm able to do 60 Percentage multiple tasking soon after i reach 100 this feature will be released untill then single task at a time"]],
    ["what is your name?", ["I'm a chatbot created by Anthropic"]],
    ["how are you(.*)", ["I'm doing good\nHow about You ?"]],
    ["sorry (.*)|sorry", ["Its alright", "Its OK, never mind"]],
    ["I am fine|good|good thanks for asking", ["Great to hear that, How can I help you?"]],
    ["i'm (.*) good|i'm good", ["Nice to hear that, How can I help you?:"]],
    ["(.*) age (.*)", ["I'm a computer program, I don't have an age!"]],
    ["what (.*) want ?|what (.*) want", ["Make me an offer I can't refuse"]],
    ["(.*) created ?", ["Anthropic created me using Python's NLTK library "]],
    ["exit", ["Goodbye!", "See you later!", "Bye!"]]
]

chat = Chat(pairs, reflections)
user_name = ""
while True:
    user_input = input("You: ")

    # Check for the exit command
    if any(keyword in user_input for keyword, _ in pairs) and user_input == "exit":
        print("Chatbot: Goodbye!")
        break
    if "my name is" in user_input.lower():
        # Extract the name from the user input
        user_name = user_input.split("my name is")[-1].strip()
        #print(f"Chatbot: Hello {user_name}, how can I assist you today?")
    elif "what is my name" in user_input.lower():
        # Check if the user has provided their name
        if user_name:
            print(f"Chatbot: Your name is {user_name}.")
            continue
        else:
            print("Chatbot: I don't know your name. Please tell me your name.")

    response = chat.respond(user_input)
    print("Chatbot:", response)


def process_user_input(user_input):
    
    stop_words = set(stopwords.words('english'))
    # Tokenize the user input
    tokens = word_tokenize(user_input.lower())
    print("tokens: ",tokens)
    # Remove punctuation and stop words
    filtered_tokens = [word for word in tokens if  word not in stop_words]
    print("filter: ",filtered_tokens)
    
    # Identify user intent, action, keywords, and book titles
    intent = None
    action = None
    keywords = []
    book_titles = []

    # Define keywords related to different intents and actions
    recommend_keywords = ["recommended","recommendations", "similar","similarly","recommend", "suggested", "books", "titles", "reading",
    "recommendations for", "book recommendations for",  "recommended books for","I need recommendations for", "book_id"
    "I need some recommendations for", "give me recommendations for","give me some recommendations for","wishlist", 
    "recommendations for", "book recommendations for","recommended books for", "similar books to", "books similar to",
    "books like","books related to", "books similar with","similar books for","similar book for","books similar as","copy",
    "books like the","books related with", "books related as","books related like",'bottom', 'book', 'favorite', 'list',"to_read","show"]
    add_keywords = ["add", "include", "insert", "append"]
    delete_keywords = ["delete", "remove", "discard", "eliminate"]
    select_keywords = ["select", "choose", "pick", "opt", "top"]

    for word in filtered_tokens:
        print(word)
        if word == "find":
            continue 
        elif word in recommend_keywords:
            keywords.append(word)
            intent = "recommend_books"
            continue
        elif word in add_keywords:
            keywords.append(word)
            action = "add"
            continue
        elif word in delete_keywords:
            keywords.append(word)
            action = "delete"
            continue
        elif word in select_keywords:
            action = "select_top"
            if "to_read" in word:
                keywords.append(word)
            else : 
                keywords.append(word)
                continue
                
        elif (word not in recommend_keywords) and (word not in add_keywords) and (word not in delete_keywords) and (word not in select_keywords):
            if action:
                # If an action keyword has been identified, consider the remaining words as keywords
                keywords.append(word)
            else:
                    # If no action keyword, consider the remaining words as potential book titles
                book_titles.append(word)   
                #print("\n",book_titles,"\n")
    #final_book_title = combine_words_to_title(book_titles)
    #rec = find_recommendations(final_book_title)
    #print("\n",rec)
    print("\n book title in pro user input ", book_titles)
    print("\n keywords in pro user input ",keywords)
    return book_titles, keywords

    

#process_user_input(input("\ninput: "))

In [351]:
rec.shape

(5, 6)

In [352]:
rec

book_id                          authors                  title  \
0     1009                     Meg Wolitzer       The Interestings   
1     4552                     Louise Penny  The Beautiful Mystery   
2     6144  Stephen Briggs, Terry Pratchett      Interesting Times   
3     9635               Jonathan Kellerman                Mystery   
4     3858                  Terry Pratchett      Interesting Times   

         isbn13  in_to_read_list  final_score  
0  9.781594e+12            False     2.946552  
1  9.781250e+12            False     2.480225  
2  9.780414e+12            False     2.289481  
3  9.780346e+12            False     2.166114  
4  9.780061e+12            False     2.148680

def respond(inputs):

    found_keywords = []
    #print(inputs)
    if (("to_read" not in inputs) and ("wishlist" not in inputs))and(("recommendations" in inputs) or ("recommend" in inputs) or ("similar" in inputs) or ("find" in inputs)) and ("favorite" not in inputs):
        found_keywords.append("recommendations")
        
    if ("favorite" in inputs) or (("recommendations" in inputs) or ("recommend" in inputs) or ("similar" in inputs) or ("find" in inputs))  and (("to_read" not in inputs) and ("wishlist" not in inputs)):
        found_keywords.append("favorite")
            
    if (("to_read" in inputs) or ("wishlist" in inputs)) or (("recommendations" in inputs) or ("recommend" in inputs) or ("similar" in inputs) or ("find" in inputs)) and ("favorite" not in inputs):
        found_keywords.append("to_read")

    if (("transfer" in inputs) or ("shift" in inputs) or ("copy" in inputs))and (("to_read" in inputs) or ("wishlist" in inputs))and ("favorite" in inputs):
        found_keywords.append("transfer")
    if not found_keywords:
        print("try again\n")
        # Default response

    if "recommendations" in found_keywords:
        return input_for_title_rec(inputs)
    
    if "favorite" in found_keywords:  
        return input_for_fav(inputs)

    if "to_read" in found_keywords:
        return input_for_to_read(inputs)
    if "transfer" in found_keywords:
        return input_for_transfer(inputs)

In [313]:
    
    # Default response
respond(input("enter\n"))

enter
find recommendations for world war

 in input_for_title_rec()
tokens:  ['find', 'recommendations', 'for', 'world', 'war']
filter:  ['find', 'recommendations', 'world', 'war']
find
recommendations
world
war

 book title in pro user input  ['world', 'war']

 keywords in pro user input  ['recommendations']
word_list: ['world', 'war'] 


combined_word: world war 

im in 
se
Found recommendations for 'world war'

    book_id                                   authors  \
0      547              H.G. Wells, Arthur C. Clarke   
1    10000                               John Keegan   
2     6564                          Sebastian Junger   
3     7305  Kimberly Brubaker Bradley, مرضیه ورشوساز   
4     3772                               Herman Wouk   

                        title        isbn13  in_to_read_list  final_score  
0       The War of the Worlds  9.780376e+12            False     3.267737  
1         The First World War  9.780376e+12            False     3.198639  
2               

In [316]:
respond(input("enter\n"))

enter
add top 1 book from the recommendations to favorite list

 input_for_fav
tokens:  ['add', 'top', '1', 'book', 'from', 'the', 'recommendations', 'to', 'favorite', 'list']
filter:  ['add', 'top', '1', 'book', 'recommendations', 'favorite', 'list']
add
top
1
book
recommendations
favorite
list

 book title in pro user input  []

 keywords in pro user input  ['add', 'top', '1', 'book', 'recommendations', 'favorite', 'list']

 book title in input for fav  []

 keywords in input for fav  ['add', 'top', '1', 'book', 'recommendations', 'favorite', 'list']

enter your user_id : 37

integers in keywords :  [1]
Entered User_Id & Book_Id not in your Favourite List, Now we are going to add into it.


In [349]:
respond(input("enter\n"))

enter
can you find recommendations for bottom book in my favorite list and show it

 input_for_fav
tokens:  ['can', 'you', 'find', 'recommendations', 'for', 'bottom', 'book', 'in', 'my', 'favorite', 'list', 'and', 'show', 'it']
filter:  ['find', 'recommendations', 'bottom', 'book', 'favorite', 'list', 'show']
find
recommendations
bottom
book
favorite
list
show

 book title in pro user input  []

 keywords in pro user input  ['recommendations', 'bottom', 'book', 'favorite', 'list', 'show']

 book title in input for fav  []

 keywords in input for fav  ['recommendations', 'bottom', 'book', 'favorite', 'list', 'show']

enter your user_id : 56

integers in keywords :  []

list of favorite in side to find direct recommendations in fav:
   Book_id                        Author  \
0    2449                   Tom Rachman   
1    3946                Karl Marlantes   
2    8498                 Laurent Binet   
3     329  Randy Pausch, Jeffrey Zaslow   
4    3253                 Claire Messud   


In [343]:
respond(input("enter\n"))

enter
add top 1 from the recommendations to favorite list and show it

 input_for_fav
tokens:  ['add', 'top', '1', 'from', 'the', 'recommendations', 'to', 'favorite', 'list', 'and', 'show', 'it']
filter:  ['add', 'top', '1', 'recommendations', 'favorite', 'list', 'show']
add
top
1
recommendations
favorite
list
show

 book title in pro user input  []

 keywords in pro user input  ['add', 'top', '1', 'recommendations', 'favorite', 'list', 'show']

 book title in input for fav  []

 keywords in input for fav  ['add', 'top', '1', 'recommendations', 'favorite', 'list', 'show']

enter your user_id : 45

integers in keywords :  [1]
Entered User_Id had this book in the Favourite List

list of favorite :
   Book_id                        Author                  Title  \
0    6836               Judith McNaught                Perfect   
1     547  H.G. Wells, Arthur C. Clarke  The War of the Worlds   

            isbn13  
0  9780743474180.0  
1  9780375759240.0  


In [365]:
respond(input("enter\n"))

enter
can you find recommendations for bottom book in my to_read list and show it

 input_for_to_read
tokens:  ['can', 'you', 'find', 'recommendations', 'for', 'bottom', 'book', 'in', 'my', 'to_read', 'list', 'and', 'show', 'it']
filter:  ['find', 'recommendations', 'bottom', 'book', 'to_read', 'list', 'show']
find
recommendations
bottom
book
to_read
list
show

 book title in pro user input  []

 keywords in pro user input  ['recommendations', 'bottom', 'book', 'to_read', 'list', 'show']

 book title in input for to_read  []

 keywords in input for fav  ['recommendations', 'bottom', 'book', 'to_read', 'list', 'show']

enter your user_id : 37

integers in keywords :  []

list of to_read in side to find direct recommendations in fav:
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1     819                                Henry David Thoreau   
2     346                        L. Frank Baum, W.W. Denslow   
3

In [366]:
respond(input("enter\n"))

enter
can you find recommendations for bottom book in my to_read list

 input_for_to_read
tokens:  ['can', 'you', 'find', 'recommendations', 'for', 'bottom', 'book', 'in', 'my', 'to_read', 'list']
filter:  ['find', 'recommendations', 'bottom', 'book', 'to_read', 'list']
find
recommendations
bottom
book
to_read
list

 book title in pro user input  []

 keywords in pro user input  ['recommendations', 'bottom', 'book', 'to_read', 'list']

 book title in input for to_read  []

 keywords in input for fav  ['recommendations', 'bottom', 'book', 'to_read', 'list']

enter your user_id : 37

integers in keywords :  []

list of to_read in side to find direct recommendations in fav:
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1     819                                Henry David Thoreau   
2     346                        L. Frank Baum, W.W. Denslow   
3    3384                                     R.L. Mathewson   

In [392]:
respond(input("\nenter\n"))


enter
add top 1 from the recommendations to to_read list and show it

 input_for_to_read
tokens:  ['add', 'top', '1', 'from', 'the', 'recommendations', 'to', 'to_read', 'list', 'and', 'show', 'it']
filter:  ['add', 'top', '1', 'recommendations', 'to_read', 'list', 'show']
add
top
1
recommendations
to_read
list
show

 book title in pro user input  []

 keywords in pro user input  ['add', 'top', '1', 'recommendations', 'to_read', 'list', 'show']

 book title in input for to_read  []

 keywords in input for fav  ['add', 'top', '1', 'recommendations', 'to_read', 'list', 'show']

enter your user_id : 37

integers in keywords :  [1]
Entered User_Id & Book_Id not in your Wishlist List, Now we are going to add into it.

list of to_read :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1     819                                Henry David Thoreau   
2     346                        L. Frank Baum, W.W. Denslow   
3 

In [435]:
respond(input("\nenter\n"))


enter
show my to_read list

 input_for_to_read
tokens:  ['show', 'my', 'to_read', 'list']
filter:  ['show', 'to_read', 'list']
show
to_read
list

 book title in pro user input  []

 keywords in pro user input  ['show', 'to_read', 'list']

 book title in input for to_read  []

 keywords in input for to_read  ['show', 'to_read', 'list']

enter your user_id : 37

integers in keywords :  []

list of to_read :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1     819                                Henry David Thoreau   
2     346                        L. Frank Baum, W.W. Denslow   
3    3384                                     R.L. Mathewson   
4    1198                                   Hanya Yanagihara   

                          Title           isbn13  
0         宮本武蔵 Miyamoto Musashi  9784770019580.0  
1  Walden; or Life in the Woods  9780691096120.0  
2    The Wonderful Wizard of Oz  9780140621680.0  


In [436]:
respond(input("\nenter\n"))


enter
delete bottom book from to_read list and show it to me

 input_for_to_read
tokens:  ['delete', 'bottom', 'book', 'from', 'to_read', 'list', 'and', 'show', 'it', 'to', 'me']
filter:  ['delete', 'bottom', 'book', 'to_read', 'list', 'show']
delete
bottom
book
to_read
list
show

 book title in pro user input  []

 keywords in pro user input  ['delete', 'bottom', 'book', 'to_read', 'list', 'show']

 book title in input for to_read  []

 keywords in input for to_read  ['delete', 'bottom', 'book', 'to_read', 'list', 'show']

enter your user_id : 37

integers in keywords :  []

delete book_id:  1198

list of favorite :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1     819                                Henry David Thoreau   
2     346                        L. Frank Baum, W.W. Denslow   
3    3384                                     R.L. Mathewson   

                          Title           isbn13  
0

In [437]:
respond(input("\nenter\n"))


enter
show my favorite list

 input_for_fav
tokens:  ['show', 'my', 'favorite', 'list']
filter:  ['show', 'favorite', 'list']
show
favorite
list

 book title in pro user input  []

 keywords in pro user input  ['show', 'favorite', 'list']

 book title in input for fav  []

 keywords in input for fav  ['show', 'favorite', 'list']

enter your user_id : 37

integers in keywords :  []

list of favorite :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1    2743                                   Noelle Stevenson   
2    4146                                     Michelle Moran   

                   Title           isbn13  
0  宮本武蔵 Miyamoto Musashi  9784770019580.0  
1                 Nimona                   
2              Nefertiti  9780307381460.0  


In [438]:
respond(input("\nenter\n"))


enter
delete bottom book from favorite list and show it to me

 input_for_fav
tokens:  ['delete', 'bottom', 'book', 'from', 'favorite', 'list', 'and', 'show', 'it', 'to', 'me']
filter:  ['delete', 'bottom', 'book', 'favorite', 'list', 'show']
delete
bottom
book
favorite
list
show

 book title in pro user input  []

 keywords in pro user input  ['delete', 'bottom', 'book', 'favorite', 'list', 'show']

 book title in input for fav  []

 keywords in input for fav  ['delete', 'bottom', 'book', 'favorite', 'list', 'show']

enter your user_id : 37

integers in keywords :  []

delete book_id:  4146

list of favorite :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1    2743                                   Noelle Stevenson   

                   Title           isbn13  
0  宮本武蔵 Miyamoto Musashi  9784770019580.0  
1                 Nimona                   


In [458]:
respond(input("\nenter\n"))


enter
add book_id 4146 to me favorite list and show it

 input_for_fav
tokens:  ['add', 'book_id', '4146', 'to', 'me', 'favorite', 'list', 'and', 'show', 'it']
filter:  ['add', 'book_id', '4146', 'favorite', 'list', 'show']
add
book_id
4146
favorite
list
show

 book title in pro user input  []

 keywords in pro user input  ['add', 'book_id', '4146', 'favorite', 'list', 'show']

 book title in input for fav  []

 keywords in input for fav  ['add', 'book_id', '4146', 'favorite', 'list', 'show']

enter your user_id : 37

integers in keywords :  [4146]
Entered User_Id & Book_Id not in your Favourite List, Now we are going to add into it.

list of favorite :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1    2743                                   Noelle Stevenson   
2    4146                                     Michelle Moran   

                   Title           isbn13  
0  宮本武蔵 Miyamoto Musashi  978477001

In [459]:
respond(input("\nenter\n"))


enter
show my favorite list

 input_for_fav
tokens:  ['show', 'my', 'favorite', 'list']
filter:  ['show', 'favorite', 'list']
show
favorite
list

 book title in pro user input  []

 keywords in pro user input  ['show', 'favorite', 'list']

 book title in input for fav  []

 keywords in input for fav  ['show', 'favorite', 'list']

enter your user_id : 37

integers in keywords :  []

list of favorite :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1    2743                                   Noelle Stevenson   
2    4146                                     Michelle Moran   

                   Title           isbn13  
0  宮本武蔵 Miyamoto Musashi  9784770019580.0  
1                 Nimona                   
2              Nefertiti  9780307381460.0  


In [460]:
respond(input("\nenter\n"))


enter
delete book_id 4146 from my favorite list and show it

 input_for_fav
tokens:  ['delete', 'book_id', '4146', 'from', 'my', 'favorite', 'list', 'and', 'show', 'it']
filter:  ['delete', 'book_id', '4146', 'favorite', 'list', 'show']
delete
book_id
4146
favorite
list
show

 book title in pro user input  []

 keywords in pro user input  ['delete', 'book_id', '4146', 'favorite', 'list', 'show']

 book title in input for fav  []

 keywords in input for fav  ['delete', 'book_id', '4146', 'favorite', 'list', 'show']

enter your user_id : 37

integers in keywords :  [4146]

list of favorite :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1    2743                                   Noelle Stevenson   

                   Title           isbn13  
0  宮本武蔵 Miyamoto Musashi  9784770019580.0  
1                 Nimona                   


In [461]:
respond(input("\nenter\n"))


enter
show me favorite list

 input_for_fav
tokens:  ['show', 'me', 'favorite', 'list']
filter:  ['show', 'favorite', 'list']
show
favorite
list

 book title in pro user input  []

 keywords in pro user input  ['show', 'favorite', 'list']

 book title in input for fav  []

 keywords in input for fav  ['show', 'favorite', 'list']

enter your user_id : 37

integers in keywords :  []

list of favorite :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1    2743                                   Noelle Stevenson   

                   Title           isbn13  
0  宮本武蔵 Miyamoto Musashi  9784770019580.0  
1                 Nimona                   


In [471]:
respond(input("\nenter\n"))


enter
show my to_read list

 input_for_to_read
tokens:  ['show', 'my', 'to_read', 'list']
filter:  ['show', 'to_read', 'list']
show
to_read
list

 book title in pro user input  []

 keywords in pro user input  ['show', 'to_read', 'list']

 book title in input for to_read  []

 keywords in input for to_read  ['show', 'to_read', 'list']

enter your user_id : 37

integers in keywords :  []

list of to_read :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1     819                                Henry David Thoreau   
2    3384                                     R.L. Mathewson   
3     346                        L. Frank Baum, W.W. Denslow   

                          Title           isbn13  
0         宮本武蔵 Miyamoto Musashi  9784770019580.0  
1  Walden; or Life in the Woods  9780691096120.0  
2                    Perfection  9780983212550.0  
3    The Wonderful Wizard of Oz  9780140621680.0  


In [472]:
respond(input("\nenter\n"))


enter
delete book_id 346 from the to_read list

 input_for_to_read
tokens:  ['delete', 'book_id', '346', 'from', 'the', 'to_read', 'list']
filter:  ['delete', 'book_id', '346', 'to_read', 'list']
delete
book_id
346
to_read
list

 book title in pro user input  []

 keywords in pro user input  ['delete', 'book_id', '346', 'to_read', 'list']

 book title in input for to_read  []

 keywords in input for to_read  ['delete', 'book_id', '346', 'to_read', 'list']

enter your user_id : 37

integers in keywords :  [346]


In [473]:
respond(input("\nenter\n"))


enter
show me to_read list

 input_for_to_read
tokens:  ['show', 'me', 'to_read', 'list']
filter:  ['show', 'to_read', 'list']
show
to_read
list

 book title in pro user input  []

 keywords in pro user input  ['show', 'to_read', 'list']

 book title in input for to_read  []

 keywords in input for to_read  ['show', 'to_read', 'list']

enter your user_id : 37

integers in keywords :  []

list of to_read :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1     819                                Henry David Thoreau   
2    3384                                     R.L. Mathewson   

                          Title           isbn13  
0         宮本武蔵 Miyamoto Musashi  9784770019580.0  
1  Walden; or Life in the Woods  9780691096120.0  
2                    Perfection  9780983212550.0  


In [474]:
respond(input("\nenter\n"))


enter
add book_id 346 to the to_read list and show it to me

 input_for_to_read
tokens:  ['add', 'book_id', '346', 'to', 'the', 'to_read', 'list', 'and', 'show', 'it', 'to', 'me']
filter:  ['add', 'book_id', '346', 'to_read', 'list', 'show']
add
book_id
346
to_read
list
show

 book title in pro user input  []

 keywords in pro user input  ['add', 'book_id', '346', 'to_read', 'list', 'show']

 book title in input for to_read  []

 keywords in input for to_read  ['add', 'book_id', '346', 'to_read', 'list', 'show']

enter your user_id : 37

integers in keywords :  [346]
Entered User_Id & Book_Id not in your Wishlist List, Now we are going to add into it.

list of favorite :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1     819                                Henry David Thoreau   
2    3384                                     R.L. Mathewson   
3     346                        L. Frank Baum, W.W. Denslow  

In [494]:
respond(input("\nenter\n"))


enter
show me to_read list

 input_for_to_read
tokens:  ['show', 'me', 'to_read', 'list']
filter:  ['show', 'to_read', 'list']
show
to_read
list

 book title in pro user input  []

 keywords in pro user input  ['show', 'to_read', 'list']

 book title in input for to_read  []

 keywords in input for to_read  ['show', 'to_read', 'list']

enter your user_id : 37

integers in keywords :  []

list of to_read :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1     819                                Henry David Thoreau   
2    3384                                     R.L. Mathewson   
3     346                        L. Frank Baum, W.W. Denslow   

                          Title           isbn13  
0         宮本武蔵 Miyamoto Musashi  9784770019580.0  
1  Walden; or Life in the Woods  9780691096120.0  
2                    Perfection  9780983212550.0  
3    The Wonderful Wizard of Oz  9780140621680.0  


In [533]:
respond(input("\nenter\n"))


enter
show me favorite list

 input_for_fav
tokens:  ['show', 'me', 'favorite', 'list']
filter:  ['show', 'favorite', 'list']
show
favorite
list

 book title in pro user input  []

 keywords in pro user input  ['show', 'favorite', 'list']

 book title in input for fav  []

 keywords in input for fav  ['show', 'favorite', 'list']

enter your user_id : 37

integers in keywords :  []

list of favorite :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1    2743                                   Noelle Stevenson   
2     346                        L. Frank Baum, W.W. Denslow   

                        Title           isbn13  
0       宮本武蔵 Miyamoto Musashi  9784770019580.0  
1                      Nimona                   
2  The Wonderful Wizard of Oz  9780140621680.0  


In [532]:
respond(input("\nenter\n"))


enter
show me to_read list

 input_for_to_read
tokens:  ['show', 'me', 'to_read', 'list']
filter:  ['show', 'to_read', 'list']
show
to_read
list

 book title in pro user input  []

 keywords in pro user input  ['show', 'to_read', 'list']

 book title in input for to_read  []

 keywords in input for to_read  ['show', 'to_read', 'list']

enter your user_id : 37

integers in keywords :  []

list of to_read :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1    3384                                     R.L. Mathewson   
2     346                        L. Frank Baum, W.W. Denslow   
3     819                                Henry David Thoreau   
4    1000                                      Leigh Bardugo   

                          Title           isbn13  
0         宮本武蔵 Miyamoto Musashi  9784770019580.0  
1                    Perfection  9780983212550.0  
2    The Wonderful Wizard of Oz  9780140621680.0  


In [527]:
respond(input("\nenter\n"))


enter
delete bottom book from favorite list and show it 

 input_for_fav
tokens:  ['delete', 'bottom', 'book', 'from', 'favorite', 'list', 'and', 'show', 'it']
filter:  ['delete', 'bottom', 'book', 'favorite', 'list', 'show']
delete
bottom
book
favorite
list
show

 book title in pro user input  []

 keywords in pro user input  ['delete', 'bottom', 'book', 'favorite', 'list', 'show']

 book title in input for fav  []

 keywords in input for fav  ['delete', 'bottom', 'book', 'favorite', 'list', 'show']

enter your user_id : 37

integers in keywords :  []

delete book_id:  1000

list of favorite :
   Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1    2743                                   Noelle Stevenson   
2     346                        L. Frank Baum, W.W. Denslow   

                        Title           isbn13  
0       宮本武蔵 Miyamoto Musashi  9784770019580.0  
1                      Nimona            

In [393]:
rec

book_id                        authors                               title  \
0     1198               Hanya Yanagihara                       A Little Life   
1     2058  Fredrik Backman, Henning Koch  Min mormor hälsar och säger förlåt   
2     3334                  Alice Hoffman           The Marriage of Opposites   
3      112                     Jojo Moyes                       Me Before You   
4     1874               Melanie Benjamin                  The Aviator's Wife   

         isbn13  in_to_read_list  final_score  
0  9.780386e+12             True     0.307764  
1  9.781501e+12             True     0.286119  
2  9.781452e+12             True     0.274038  
3  9.780670e+12             True     0.263627  
4  9.780346e+12             True     0.262896

def input_for_fav(inputs):
    global rec
    print("\n input_for_fav")
    book_title, keywords = process_user_input(inputs)
    print("\n book title in input for fav ", book_title)
    print("\n keywords in input for fav ",keywords)
    user_id = input("\nenter your user_id : ")
    #my_list = ["hello", "top", "1", "cat", "10"]
    int_vals = []
    for item in keywords:
        try: 
            int_val = int(item)
            int_vals.append(int_val)
        except ValueError:
            pass 
    print("\nintegers in keywords : ", int_vals)
    if (("top" in keywords) or ("bottom" in keywords)) and ("favorite" in keywords) and (("recommendations" in keywords) or ("similar" in keywords) or ("related" in keywords)) and ("add" not in keywords):
        #user_id = input("\nenter your user_id : ")
        df = direct_input_fav(user_id)
        print("\nlist of favorite in side to find direct recommendations in fav:\n", df)
        if ("top" in keywords): 
            row = int_vals[0]
            rec = find_recommendations(df.iloc[row-1][2])
            print("\n",rec)
        if ("bottom" in keywords): 
            int_vals.append(1)
            row = int_vals[0]
            print("\nrow: ",row)
            rec = find_recommendations(df.iloc[-row][2])
            print("\n",rec)
        '''if "show" in keywords:
            df = direct_input_fav(user_id)
            print("\nlist of favorite :\n", df)'''
    if len(book_title) > 0 :
        final_book_title = combine_words_to_title(book_title)
        rec = find_recommendations(final_book_title)
        print("\n",rec)
            
    if ("add" in keywords) and (("top" in keywords) or ("bottom" in keywords)) and ("favorite" in keywords) :
        #user_id = input("\nenter your user_id : ")
        if "top" in keywords:
            row = int_vals[0]
            add_fav(rec.iloc[row-1][0],str(user_id))
        if "bottom" in keywords:
            int_vals.append(1)
            row = int_vals[0]
            add_fav(rec.iloc[-row][0], str(user_id))
        if "show" in keywords:
            df = direct_input_fav(user_id)
            print("\nlist of favorite :\n", df)
            
    if ("book_id" in keywords and "add" in keywords and "favorite" in keywords):
        add_fav(int_vals[0],str(user_id))
        if "show" in keywords:
            df = direct_input_fav(user_id)
            print("\nlist of favorite :\n", df)
            
    if ("book_id" in keywords and "delete" in keywords and "favorite" in keywords):
        delete_from_fav(str(int_vals[0]),str(user_id))
        if "show" in keywords:
            df = direct_input_fav(user_id)
            print("\nlist of favorite :\n", df)
        
    if ("delete" in keywords) and (("top" in keywords) or ("bottom" in keywords)) and ("favorite" in keywords) :
        #user_id = input("\nenter your user_id : ")
        df = direct_input_fav(user_id)
        if "top" in keywords:
            row = int_vals[0]
            delete_from_fav(df.iloc[row-1][0],str(user_id))
        elif "bottom" in keywords:
            int_vals.append(1)
            row = int_vals[0]
            print("\ndelete book_id: ", df.iloc[-row][0])
            delete_from_fav(df.iloc[-row][0], str(user_id))
        if "show" in keywords:
            df = direct_input_fav(user_id)
            print("\nlist of favorite :\n", df)
    if ("show" in keywords) and ("favorite" in keywords) and (("recommendations" not in keywords) or ("similar" not in keywords) or ("related" not in keywords)) and ("add" not in keywords) and ('delete' not in keywords):
        df = direct_input_fav(user_id)
        print("\nlist of favorite :\n", df)

def input_for_to_read(inputs):
    global rec
    print("\n input_for_to_read")
    book_title, keywords = process_user_input(inputs)
    print("\n book title in input for to_read ", book_title)
    print("\n keywords in input for to_read ",keywords)
    user_id = input("\nenter your user_id : ")
    #my_list = ["hello", "top", "1", "cat", "10"]
    int_vals = []
    for item in keywords:
        try: 
            int_val = int(item)
            int_vals.append(int_val)
        except ValueError:
            pass 
    print("\nintegers in keywords : ", int_vals)
    if (("top" in keywords) or ("bottom" in keywords)) and (("to_read" in keywords)or ("wishlist" in keywords)) and (("recommendations" in keywords) or ("similar" in keywords) or ("related" in keywords)) and ("add" not in keywords):
        #user_id = input("\nenter your user_id : ")
        df = direct_input_to_read(user_id)
        print("\nlist of to_read in side to find direct recommendations in to_read:\n", df)
        if ("top" in keywords): 
            row = int_vals[0]
            rec = find_recommendations(df.iloc[row-1][2])
            print("\n",rec)
        if ("bottom" in keywords): 
            int_vals.append(1)
            row = int_vals[0]
            print("\nrow: ",row)
            rec = find_recommendations(df.iloc[-row][2])
            print("\n",rec)
        '''if "show" in keywords:
            df = direct_input_fav(user_id)
            print("\nlist of to_read :\n", df)'''
    if len(book_title) > 0 :
        final_book_title = combine_words_to_title(book_title)
        rec = find_recommendations(final_book_title)
        print("\n",rec)

    if ("add" in keywords) and (("top" in keywords) or ("bottom" in keywords)) and (("to_read" in keywords)or ("wishlist" in keywords)) :
        #user_id = input("\nenter your user_id : ")
        if "top" in keywords:
            row = int_vals[0]
            add_to_read(rec.iloc[row-1][0],str(user_id))
        if "bottom" in keywords:
            int_vals.append(1)
            row = int_vals[0]
            add_to_read(rec.iloc[-row][0], str(user_id))
        if "show" in keywords:
            df = direct_input_to_read(user_id)
            print("\nlist of to_read :\n", df)
    if ("book_id" in keywords and "add" in keywords and (("to_read" in keywords)or ("wishlist" in keywords))):
        add_to_read(int_vals[0],str(user_id))
        if "show" in keywords:
            df = direct_input_to_read(user_id)
            print("\nlist of to_read :\n", df)
            
    if ("book_id" in keywords and "delete" in keywords and (("to_read" in keywords)or ("wishlist" in keywords))):
        delete_from_to_read(str(int_vals[0]),str(user_id))
        if "show" in keywords:
            df = direct_input_to_read(user_id)
            print("\nlist of to_read :\n", df)
    if ("delete" in keywords) and (("top" in keywords) or ("bottom" in keywords)) and (("to_read" in keywords)or ("wishlist" in keywords)) :
        #user_id = input("\nenter your user_id : ")
        df = direct_input_to_read(user_id)
        if "book_id" in keywords:
            delete_from_to_read(int_vals[0],str(user_id))
        else:
            if "top" in keywords:
                row = int_vals[0]
                delete_from_to_read(df.iloc[row-1][0],str(user_id))
            elif "bottom" in keywords:
                int_vals.append(1)
                row = int_vals[0]
                print("\ndelete book_id: ", df.iloc[-row][0])
                delete_from_to_read(df.iloc[-row][0], str(user_id))
        if "show" in keywords:
            df = direct_input_to_read(user_id)
            print("\nlist of to_read :\n", df)
    if ("show" in keywords) and (("to_read" in keywords) or ("wishlist" in keywords)) and (("recommendations" not in keywords) or ("similar" not in keywords) or ("related" not in keywords)) and ("add" not in keywords) and ('delete' not in keywords):
        df = direct_input_to_read(user_id)
        print("\nlist of to_read :\n", df)

        
        
        
        
        

In [ ]:
def input_for_transfer(from_file, to_file,book_id,user_id):
    #we have code for this see below
    print(from_file, to_file,book_id,user_id)
    print("\n transfer")
    if from_file == "to_read":
        transfer_to_fav_from_to_read(book_id,user_id)
        print("\ni side fav from file")
    if from_file == 'favorite':
        transfer_to_to_read_from_fav(book_id, user_id)
        print("\nin side to_read from file")

In [236]:
rec.iloc[1][0]


4146

In [235]:
rec

book_id                              authors               title  \
0     7689  Andrew Stanton, Walt Disney Company        Finding Nemo   
1     4146                       Michelle Moran           Nefertiti   
2     2743                     Noelle Stevenson              Nimona   
3     8520            Ayşe Kulin, John W. Baker        Nefes Nefese   
4     3978                         Isaac Asimov  The Robots of Dawn   

         isbn13  in_to_read_list  final_score  
0  9.781592e+12            False     3.748958  
1  9.780307e+12            False     1.845089  
2           NaN            False     1.666022  
3           NaN            False     1.603502  
4  9.780553e+12            False     1.525459

In [317]:
df = direct_input_fav(37)

In [318]:
df

Book_id                                             Author  \
0    7173  Eiji Yoshikawa, Charles Terry, Edwin O. Reisch...   
1    2743                                   Noelle Stevenson   
2    4146                                     Michelle Moran   
3    3978                                       Isaac Asimov   
4     547                       H.G. Wells, Arthur C. Clarke   

                   Title           isbn13  
0  宮本武蔵 Miyamoto Musashi  9784770019580.0  
1                 Nimona                   
2              Nefertiti  9780307381460.0  
3     The Robots of Dawn  9780553299500.0  
4  The War of the Worlds  9780375759240.0

In [289]:
df.iloc[1][2]

'Nimona'

In [545]:
import re
import csv

list_to_file = {
    "to_read list": 'to_read',
    "favorite list": 'favorite'
}


def process_user_input_for_transfer(user_input):

    book_id_match = re.search(r"book_id (\d+)", user_input)
    if book_id_match:  
        book_id = book_id_match.group(1)
    else:
        print("Book id not found")
        return

    from_list_match = re.search(r"from (\w+\s?list)", user_input)
    if from_list_match:
        from_list = from_list_match.group(1)
        from_file = list_to_file[from_list]
    else:
        print("Source file not found")
        return

    to_list_match = re.search(r"to (\w+\s?list)", user_input)  
    if to_list_match:   
        to_list = to_list_match.group(1)
        to_file = list_to_file[to_list]
    else:
        print("Destination file not found")
        return

    print(f"Copying book {book_id} from {from_file} to {to_file}")
    user_id=input("\nenter user_id : ")
    input_for_transfer(from_file, to_file,book_id,user_id)
    #copy_row(from_file, to_file, book_id,user_id)

user_input = "copy book_id 1000 from to_read list to favorite list and show it"
#user_input = "copy book_id 2516 from to_read to favorite list"
process_user_input_for_transfer(user_input)

Copying book 1000 from to_read to favorite

enter user_id : 567
to_read favorite 1000 567

 transfer
Entered User_Id had this book in the to_read List
Entered User_Id & Book_Id not in your Favourite List, Now we are going to add into it.

i side fav from file


def process_user_input_title_extraction(self,user_input):
    import nltk
    from nltk.tokenize import word_tokenize
        from nltk.corpus import stopwords
        from nltk import pos_tag, ne_chunk

        #nltk.download('punkt')
        #nltk.download('stopwords')
        #nltk.download('maxent_ne_chunker')
        #nltk.download('words')

        stop_words = set(stopwords.words('english'))
        # Tokenize the user input
        tokens = word_tokenize(user_input.lower())
        print("tokens: ",tokens)
        # Remove punctuation and stop words
        filtered_tokens = [word for word in tokens if  word not in stop_words]
        print("filter: ",filtered_tokens)

        # Identify user intent, action, keywords, and book titles
        intent = None
        action = None
        keywords = []
        book_titles = []

        # Define keywords related to different intents and actions
        recommend_keywords = ["recommended","recommendations", "recommend", "suggested", "books", "titles", "reading","recommendations for", "book recommendations for",  "recommended books for",
        "I need recommendations for", "I need some recommendations for", "give me recommendations for",
        "give me some recommendations for", "recommendations for", "book recommendations for","recommended books for", 
        "similar books to", "books similar to", "books like","books related to", "books similar with","similar books for",
        "books similar as","books like the","books related with", "books related as","books related like"]
        add_keywords = ["add", "include", "insert", "append"]
        delete_keywords = ["delete", "remove", "discard", "eliminate"]
        select_keywords = ["select", "choose", "pick", "opt", "top"]

        for word in filtered_tokens:
            print(word)
            if word in recommend_keywords:
                intent = "recommend_books"
            elif word in add_keywords:
                action = "add"
            elif word in delete_keywords:
                action = "delete"
            elif word in select_keywords:
                action = "select_top"
                if "to_read" in word:
                    keywords.append(word)
            else:
                if action:
                    # If an action keyword has been identified, consider the remaining words as keywords
                    keywords.append(word)
                else:
                    # If no action keyword, consider the remaining words as potential book titles
                    book_titles.append(word)        
        final_book_title = self.combine_words_to_title(book_titles)
        #return final_book_title


def direct_search_fav(book_id, user_id):
    filename = '/Users/cdmstudent/Desktop/DSC-540 Adv ML/Final Project/ChatBot/PersonalizedBookSearch/goodbooks-10k/favourite.csv'
    with open(filename, mode='r') as file:
    reader = csv.reader(file)
    #book_id = input("enter book_id: ")
    #user_id = input("enter user_id: ")
    i =1
    for row in reader:
        if user_id in row and book_id in row:
            print("Entered User_Id had this book in the Favourite List")
            i = 0
            break
        else: continue

def direct_search_to_read():
    filename = '/Users/cdmstudent/Desktop/DSC-540 Adv ML/Final Project/ChatBot/PersonalizedBookSearch/goodbooks-10k/to_read.csv'
    with open(filename, mode='r') as file:
    reader = csv.reader(file)
    book_id = input("enter book_id: ")
    user_id = input("enter user_id: ")
    i =1
    for row in reader:
        if user_id in row and book_id in row:
            print("Entered User_Id had this book in the to_read List")
            i = 0
            break
        else: continue

def find_recommendations(title):
    print("im in ")
    #import SearchEngine as SE
    se = SE.SearchEngine()
    print("se")
    # Logic to fetch recommendations based on the title (replace this with your actual recommendation logic)
    recommendations = se.query(text=title,user_id= 1,  to_read_boosting_factor=1.5, K=5)
    #print(recommendations)
    print("Found recommendations for '{}'".format(title))
    return recommendations

def input_for_title_rec(inputs):
    global rec
    print("\n in input_for_title_rec()")
    book_title, keywords = process_user_input(inputs)
    final_book_title = combine_words_to_title(book_title)
    rec = find_recommendations(final_book_title)
    print("\n",rec)

def combine_words_to_title(word_list):
    print("word_list:",word_list,"\n")
    keywords_to_drop = ["need", "title","i","book","name","find",'related', 'like',"give","recommendations","related","similar"]
    # Remove the selected keywords from the word list
    filtered_words = [word for word in word_list if word not in keywords_to_drop]

    # Combine the remaining words into a single string
    combined_word = ' '.join(filtered_words)
    print("\ncombined_word:",combined_word,"\n")
    #find_recommendations(combined_word)
    return combined_word


def direct_input_to_read(user_id):
    with open('/Users/cdmstudent/Desktop/DSC-540 Adv ML/Final Project/ChatBot/PersonalizedBookSearch/goodbooks-10k/to_read.csv', 'r') as file:

        # create a CSV reader object
        reader = csv.reader(file)

        # read the header row
        header = next(reader)

        # create an empty list to store matching rows
        matching_rows = []

        # get the user input
        entered_value = str(user_id) #input('Enter User_id to search:: ')

        # loop through the rows and check for matches
        for row in reader:
            if entered_value == row[0]:

                with open('/Users/cdmstudent/Desktop/DSC-540 Adv ML/Final Project/ChatBot/PersonalizedBookSearch/goodbooks-10k/books_cleaned.csv', 'r') as file1:
                    reader1 = csv.reader(file1)
                    header = next(reader1)
                    for row1 in reader1:
                        if row[1] == row1[0]:
                            matching_rows.append(row1)

        df = pd.DataFrame(matching_rows, columns = ['Book_id','Author','Title','isbn13'])
        #print(df)
        return df

def direct_input_fav(user_id):
# open the CSV file
    with open('/Users/cdmstudent/Desktop/DSC-540 Adv ML/Final Project/ChatBot/PersonalizedBookSearch/goodbooks-10k/favourite.csv', 'r') as file:

        # create a CSV reader object
        reader = csv.reader(file)

        # read the header row
        header = next(reader)

        # create an empty list to store matching rows
        matching_rows = []

        # get the user input
        entered_value = str(user_id) #input('Enter User_Id to search: ')

        # loop through the rows and check for matches
        for row in reader:
            if entered_value == row[0]:
                with open('/Users/cdmstudent/Desktop/DSC-540 Adv ML/Final Project/ChatBot/PersonalizedBookSearch/goodbooks-10k/books_cleaned.csv', 'r') as file1:
                    reader1 = csv.reader(file1)
                    header = next(reader1)
                    for row1 in reader1:
                        if row[1] == row1[0]:
                            matching_rows.append(row1)

        df = pd.DataFrame(matching_rows, columns = ['Book_id','Author','Title','isbn13'])
        #print(df)
        return df

def add_to_read(book_id,user_id):
    filename = '/Users/cdmstudent/Desktop/DSC-540 Adv ML/Final Project/ChatBot/PersonalizedBookSearch/goodbooks-10k/to_read.csv'

    # Create a new CSV file with headers for two columns User_Id & Book_Id if it doesn't exist
    if not os.path.isfile(filename):
        with open(filename, mode='w', newline='') as file:
            writer = csv.writer(file)
            writer.writerow(['User_Id', 'Book_Id'])  # Write headers for the two columns

    with open(filename, mode='r') as file:
        reader = csv.reader(file)
        #book_id = input("enter book_id: ")
        #user_id = input("enter user_id: ")
        i =1
        for row in reader:
            if str(user_id) in row and str(book_id) in row:
                print("Entered User_Id had this book in the to_read List")
                i = 0
                break
            else: continue
    if(i==1):
        # Open the CSV file in append mode
        with open(filename, mode='a', newline='') as file:
            print("Entered User_Id & Book_Id not in your Wishlist List, Now we are going to add into it.") 
            writer = csv.writer(file)
            new_row = [user_id,book_id]

            # Write the new row to the CSV file
            writer.writerow(new_row)
    return

def add_fav(book_id,user_id):
    filename = '/Users/cdmstudent/Desktop/DSC-540 Adv ML/Final Project/ChatBot/PersonalizedBookSearch/goodbooks-10k/favourite.csv'

    # Create a new CSV file with headers for two columns User_Id & Book_Id if it doesn't exist
    if not os.path.isfile(filename):
        with open(filename, mode='w', newline='') as file:
            writer = csv.writer(file)
            writer.writerow(['User_Id', 'Book_Id'])  # Write headers for the two columns

    with open(filename, mode='r') as file:
        reader = csv.reader(file)
        #book_id = input("enter book_id: ")
        #user_id = input("enter user_id: ")
        i =1
        for row in reader:
            if str(user_id) in row and str(book_id) in row:
                print("Entered User_Id had this book in the Favourite List")
                i = 0
                break
            else: continue
    if(i==1):
        # Open the CSV file in append mode
        with open(filename, mode='a', newline='') as file:
            print("Entered User_Id & Book_Id not in your Favourite List, Now we are going to add into it.") 
            writer = csv.writer(file)
            new_row = [user_id,book_id]

            # Write the new row to the CSV file
            writer.writerow(new_row)
    return

def delete_from_fav(book_id,user_id):
    filename = '/Users/cdmstudent/Desktop/DSC-540 Adv ML/Final Project/ChatBot/PersonalizedBookSearch/goodbooks-10k/favourite.csv'

    # Read the existing data from the CSV file
    with open(filename, mode='r', newline='') as file:
        reader = csv.reader(file)
        data = list(reader)

    # Find and remove the specified book ID for the given user ID
    updated_data = [row for row in data if row != [user_id, book_id]]

    # Write the updated data back to the CSV file
    with open(filename, mode='w', newline='') as file:
        writer = csv.writer(file)
        writer.writerows(updated_data)
    return

# Example usage:
#user_id = "37"
#book_id = "4146"
#user_id = input("user_id")
#book_id = input("book_id")
#delete_from_fav(user_id, book_id)


def delete_from_to_read(book_id,user_id):
    filename = '/Users/cdmstudent/Desktop/DSC-540 Adv ML/Final Project/ChatBot/PersonalizedBookSearch/goodbooks-10k/to_read.csv'

    # Read the existing data from the CSV file
    with open(filename, mode='r', newline='') as file:
        reader = csv.reader(file)
        data = list(reader)

    # Find and remove the specified book ID for the given user ID
    updated_data = [row for row in data if row != [user_id, book_id]]

    # Write the updated data back to the CSV file
    with open(filename, mode='w', newline='') as file:
        writer = csv.writer(file)
        writer.writerows(updated_data)
    return
# Example usage:
#user_id = "123"
#book_id = "456"

#delete_from_fav(user_id, book_id)


In [529]:
def transfer_to_fav_from_to_read(book_id, user_id):
    filename = '/Users/cdmstudent/Desktop/DSC-540 Adv ML/Final Project/ChatBot/PersonalizedBookSearch/goodbooks-10k/to_read.csv'
    with open(filename, mode='r') as file:
        reader = csv.reader(file)
        #book_id = input("enter book_id: ")
        #user_id = input("enter user_id: ")
        i =1
        for row in reader:
            if user_id in row and book_id in row:
                print("Entered User_Id had this book in the to_read List")
                i = 0
                break
            else: continue
    filename = '/Users/cdmstudent/Desktop/DSC-540 Adv ML/Final Project/ChatBot/PersonalizedBookSearch/goodbooks-10k/favourite.csv'
    # Open the CSV file in append mode
    with open(filename, mode='a', newline='') as file:
        print("Entered User_Id & Book_Id not in your Favourite List, Now we are going to add into it.") 
        writer = csv.writer(file)
        new_row = [user_id,book_id]

        # Write the new row to the CSV file
        writer.writerow(new_row)
    return

In [528]:
def transfer_to_to_read_from_fav(book_id, user_id):
    filename = '/Users/cdmstudent/Desktop/DSC-540 Adv ML/Final Project/ChatBot/PersonalizedBookSearch/goodbooks-10k/favourite.csv'
    with open(filename, mode='r') as file:
        reader = csv.reader(file)
        #book_id = input("enter book_id: ")
        #user_id = input("enter user_id: ")
        i =1
        for row in reader:
            if user_id in row and book_id in row:
                print("Entered User_Id had this book in the favourite List")
                i = 0
                break
            else: continue
    filename = '/Users/cdmstudent/Desktop/DSC-540 Adv ML/Final Project/ChatBot/PersonalizedBookSearch/goodbooks-10k/to_read.csv'
    # Open the CSV file in append mode
    with open(filename, mode='a', newline='') as file:
        print("Entered User_Id & Book_Id not in your to_read List, Now we are going to add into it.") 
        writer = csv.writer(file)
        new_row = [user_id,book_id]

        # Write the new row to the CSV file
        writer.writerow(new_row)
    return